server/src/mysql/worker.rs

```diff
@@ -0,0 +1,132 @@
+use query_engine::executor::Executor as QueryExecutor;
+use snafu::ResultExt;
+use table_engine::engine::EngineRuntimes;
+
+use crate::{
+    context::RequestContext,
+    handlers::{self, sql::Request},
+    instance::Instance,
+    mysql::{error::*, writer::MysqlQueryResultWriter},
+};
+
+pub struct MysqlWorker<W: std::io::Write + Send + Sync, C, Q> {
+    generic_hold: PhantomData<W>,
+    instance: Arc<Instance<C, Q>>,
+    runtimes: Arc<EngineRuntimes>,
+}
+
+impl<W: std::io::Write + Send + Sync, C: CatalogManager + 'static, Q: QueryExecutor + 'static>
+    MysqlWorker<W, C, Q>
+{
+    pub fn new(instance: Arc<Instance<C, Q>>, runtimes: Arc<EngineRuntimes>) -> Self {
+        Self {
+            generic_hold: PhantomData::default(),
+            instance,
+            runtimes,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<W: std::io::Write + Send + Sync, C: CatalogManager + 'static, Q: QueryExecutor + 'static>
+    AsyncMysqlShim<W> for MysqlWorker<W, C, Q>
+{
+    type Error = crate::mysql::error::Error;
+
+    async fn on_prepare<'a>(
+        &'a mut self,
+        _query: &'a str,
+        info: StatementMetaWriter<'a, W>,
+    ) -> Result<()> {
+        info.error(
+            ErrorKind::ER_UNKNOWN_ERROR,
+            "Prepare is not supported in CeresDB".as_bytes(),
+        )?;
+        Ok(())
+    }
+
+    async fn on_execute<'a>(
+        &'a mut self,
+        _id: u32,
+        _param: opensrv_mysql::ParamParser<'a>,
+        writer: QueryResultWriter<'a, W>,
+    ) -> Result<()> {
+        writer.error(
+            ErrorKind::ER_UNKNOWN_ERROR,
+            "Execute is not supported in CeresDB".as_bytes(),
+        )?;
+        Ok(())
+    }
+
+    async fn on_close(&mut self, id: u32) {
+        info!("client(id={}) wishes to deallocate resources associated with a previously prepared statement.", id)
+    }
+
+    async fn on_query<'a>(
+        &'a mut self,
+        sql: &'a str,
+        results: QueryResultWriter<'a, W>,
+    ) -> Result<()> {
+        let ctx = match self.create_ctx() {
+            Ok(res) => res,
+            Err(error) => return Self::err(results, error),
+        };
+
+        let req = Request::from(sql.to_string());
+        let result = handlers::sql::handle_sql(ctx, self.instance.clone(), req)
+            .await
+            .map_err(|e| {
+                error!("Mysql service Failed to handle sql, err: {}", e);
+                e
+            })
+            .context(HandleSql {
+                sql: sql.to_string(),
+            });
+        match result {
+            Ok(res) => {
+                let mut writer = MysqlQueryResultWriter::create(results);
+                writer.write(res)
+            }
+            Err(err) => return Self::err(results, err),
+        }
+    }
+}
+
+impl<W: std::io::Write + Sync + Send, C: CatalogManager + 'static, Q: QueryExecutor + 'static>
+    MysqlWorker<W, C, Q>
+{
+    fn err(writer: QueryResultWriter<'_, W>, error: Error) -> Result<()> {
```

Comment:
It seems this method is useless now because the `Error` type is already `crate::mysql::error::Error` and we can just use `?` to return the error directly.

Reply:
For example, the codes (L77-L80):
```rust
let ctx = match self.create_ctx() {
    Ok(res) => res,
    Err(error) => return Self::err(results, error),
};
```

can written as:
```rust
let ctx = self.create_ctx()?;
```

Reply:
I'm sorry that I misunderstand the usage of the `QueryResultWriter`. 

So the error must be returned by `writer.error` method? 

If it does, I guess we can define an method called `do_query( &'a mut self, sql: &'a str)` which just returns the query result and then in `on_query` use the `QueryResultWriter` to process the query result from the `do_query`. After that, I guess the error handling will be more concise and elegant.